src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

```diff
@@ -1480,15 +1473,32 @@ private static object CreateCaObject(RuntimeModule module, RuntimeType type, IRu
             blobStart = (IntPtr)pBlobStart;
         }
 
-        private static object[] CreateAttributeArrayHelper(RuntimeType elementType, int elementCount)
+        private static object[] CreateAttributeArrayHelper(RuntimeType caType, int elementCount)
         {
+            if (caType.IsValueType)
+            {
+                return CreateArray<object>();
+            }
+
+            if (caType.ContainsGenericParameters)
+            {
+                if (caType.IsSubclassOf(typeof(Attribute)))
+                {
+                    return CreateArray<Attribute>();
+                }
+
+                return CreateArray<object>();
+            }
+
             // If we have 0 elements, don't allocate a new array
             if (elementCount == 0)
             {
-                return elementType.GetEmptyArray();
+                return caType.GetEmptyArray();
             }
 
-            return (object[])Array.CreateInstance(elementType, elementCount);
+            return (object[])Array.CreateInstance(caType, elementCount);
+
+            T[] CreateArray<T>() => elementCount == 0 ? Array.Empty<T>() : new T[elementCount];
```

Comment:
(Non-static) local methods with captured arguments are inefficient. The compiler will create display type to pass the captured locals around that comes with a lot of ceremony. 

I think I would inline it. The local method is not worth it. Also, I think it may be worth it to check for Attribute and use the faster non-reflection based path for it. Something like:

```
bool useAttributeArray = false;
bool useObjectArray = false;

if (caType == typeof(Attribute))
{
    useAttributeArray = true;
}
else if (caType.IsValueType)
{
    useObjectArray = true;
}
else if (caType.ContainsGenericParameters)
{
    if (caType.IsSubclassOf(typeof(Attribute)))
    {
        useAttributeArray = true;
    }
    else
    {
        useObjectArray = true;
    } 
}

if (useAttributeArray) (elementCount != 0) ? new Attribute[elementCount] :  return Array.Empty<Attribute>();
if (useObjectArray) (elementCount != 0) ? new object[elementCount] : Array.Empty<object>();
return (elementCount != 0) ? (object[])Array.CreateInstance(caType, elementCount) : caType.GetEmptyArray();
```